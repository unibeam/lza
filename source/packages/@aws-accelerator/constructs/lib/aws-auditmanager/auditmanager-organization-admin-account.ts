/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Initialized AuditManagerOrganizationalAdminAccountProps properties
 */
export interface AuditManagerOrganizationalAdminAccountProps {
  /**
   * Admin account id
   */
  readonly adminAccountId: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 * Class for AuditManagerOrganizationAdminAccount
 */
export class AuditManagerOrganizationAdminAccount extends Construct {
  public readonly id: string;

  constructor(scope: Construct, id: string, props: AuditManagerOrganizationalAdminAccountProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::AuditManagerEnableOrganizationAdminAccount';

    const commonPolicyStatements = [
      {
        Sid: 'AuditManagerEnableOrganizationAdminAccountTaskOrganizationActions',
        Effect: 'Allow',
        Action: [
          'organizations:DeregisterDelegatedAdministrator',
          'organizations:DescribeOrganization',
          'organizations:EnableAWSServiceAccess',
          'organizations:ListAWSServiceAccessForOrganization',
          'organizations:ListAccounts',
          'organizations:ListDelegatedAdministrators',
          'organizations:RegisterDelegatedAdministrator',
        ],
        Resource: '*',
        Condition: {
          StringLikeIfExists: {
            'organizations:DeregisterDelegatedAdministrator': ['auditmanager.amazonaws.com'],
            'organizations:DescribeOrganization': ['auditmanager.amazonaws.com'],
            'organizations:EnableAWSServiceAccess': ['auditmanager.amazonaws.com'],
            'organizations:ListAWSServiceAccessForOrganization': ['auditmanager.amazonaws.com'],
            'organizations:ListAccounts': ['auditmanager.amazonaws.com'],
            'organizations:ListDelegatedAdministrators': ['auditmanager.amazonaws.com'],
            'organizations:RegisterDelegatedAdministrator': ['auditmanager.amazonaws.com'],
          },
        },
      },
      {
        Sid: 'AuditManagerEnableOrganizationAdminAccountTaskDetectiveActions',
        Effect: 'Allow',
        Action: [
          'auditmanager:RegisterAccount',
          'auditmanager:DeregisterAccount',
          'auditmanager:RegisterOrganizationAdminAccount',
          'auditmanager:DeregisterOrganizationAdminAccount',
          'auditmanager:getOrganizationAdminAccount',
        ],
        Resource: '*',
      },
      {
        Sid: 'ServiceLinkedRoleDetective',
        Effect: 'Allow',
        Action: ['iam:CreateServiceLinkedRole'],
        Resource: ['*'],
      },
    ];

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'enable-organization-admin-account/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: props.kmsKey
        ? [
            ...commonPolicyStatements,
            {
              Sid: 'AuditManagerEnableKmsKeyGrants',
              Effect: 'Allow',
              Action: 'kms:CreateGrant',
              Resource: props.kmsKey.keyArn,
              Condition: {
                StringLike: {
                  'kms:ViaService': 'auditmanager.*.amazonaws.com',
                },
                Bool: {
                  'kms:GrantIsForAWSResource': 'true',
                },
              },
            },
          ]
        : [...commonPolicyStatements],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        adminAccountId: props.adminAccountId,
        kmsKeyArn: props.kmsKey ? props.kmsKey.keyArn : undefined,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
