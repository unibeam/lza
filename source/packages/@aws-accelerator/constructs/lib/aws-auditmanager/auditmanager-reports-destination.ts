/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

const path = require('path');

/**
 * Export config destination types
 */
export enum AuditManagerDefaultReportsDestinationTypes {
  S3 = 'S3',
}

/**
 * Initialized AuditManagerDefaultReportDestinationProps properties
 */
export interface AuditManagerDefaultReportsDestinationProps {
  /**
   * Export destination type
   */
  readonly defaultReportsDestinationType: string;
  /**
   * Publishing destination bucket arn
   */
  readonly bucket: string;
  /**
   * Publishing destination bucket KMS key
   */
  readonly bucketKmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly kmsKey?: cdk.aws_kms.IKey;
  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

/**
 * Class - AuditManagerDefaultsReportDestination
 */
export class AuditManagerDefaultReportsDestination extends Construct {
  public readonly id: string = '';

  constructor(scope: Construct, id: string, props: AuditManagerDefaultReportsDestinationProps) {
    super(scope, id);

    const RESOURCE_TYPE = 'Custom::AuditManagerCreateDefaultReportsDestination';

    const provider = cdk.CustomResourceProvider.getOrCreateProvider(this, RESOURCE_TYPE, {
      codeDirectory: path.join(__dirname, 'create-reports-destination/dist'),
      runtime: cdk.CustomResourceProviderRuntime.NODEJS_18_X,
      policyStatements: [
        {
          Sid: 'AuditManagerCreatePublishingDestinationCommandTaskAuditManagerActions',
          Effect: 'Allow',
          Action: ['auditmanager:UpdateSettings'],
          Resource: '*',
        },
      ],
    });

    const resource = new cdk.CustomResource(this, 'Resource', {
      resourceType: RESOURCE_TYPE,
      serviceToken: provider.serviceToken,
      properties: {
        region: cdk.Stack.of(this).region,
        defaultReportsDestinationType: props.defaultReportsDestinationType,
        bucket: props.bucket,
        kmsKeyArn: props.bucketKmsKey ? props.bucketKmsKey.keyArn : undefined,
      },
    });

    /**
     * Singleton pattern to define the log group for the singleton function
     * in the stack
     */
    const stack = cdk.Stack.of(scope);
    const logGroup =
      (stack.node.tryFindChild(`${provider.node.id}LogGroup`) as cdk.aws_logs.LogGroup) ??
      new cdk.aws_logs.LogGroup(stack, `${provider.node.id}LogGroup`, {
        logGroupName: `/aws/lambda/${(provider.node.findChild('Handler') as cdk.aws_lambda.CfnFunction).ref}`,
        retention: props.logRetentionInDays,
        encryptionKey: props.kmsKey,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
      });
    resource.node.addDependency(logGroup);

    this.id = resource.ref;
  }
}
