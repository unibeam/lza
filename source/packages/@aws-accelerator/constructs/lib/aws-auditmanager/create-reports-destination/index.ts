/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * create-auditmanager-default-reports-destination - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const region = event.ResourceProperties['region'];
  const defaultReportsDestinationType = event.ResourceProperties['defaultReportsDestinationType'];
  const bucket = event.ResourceProperties['bucket'];
  const kmsKeyArn: string | undefined = event.ResourceProperties['kmsKeyArn'] ?? undefined;
  const solutionId = process.env['SOLUTION_ID'];

  const auditManagerClient = new AWS.AuditManager({ region: region, customUserAgent: solutionId });

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      console.log('starting - CreateDefaultLoggingDestination');

      await throttlingBackOff(() =>
        auditManagerClient
          .updateSettings({
            defaultAssessmentReportsDestination: {
              destination: bucket,
              destinationType: defaultReportsDestinationType,
            },
            kmsKey: kmsKeyArn,
          })
          .promise(),
      );

      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      console.log('starting - DeleteDefaultReportDestination');

      return { Status: 'Success', StatusCode: 200 };
  }
}
