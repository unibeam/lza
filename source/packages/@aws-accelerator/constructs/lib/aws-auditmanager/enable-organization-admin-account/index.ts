/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import * as AWS from 'aws-sdk';
AWS.config.logger = console;

/**
 * enable-auditmanager - lambda handler
 *
 * @param event
 * @returns
 */
export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      Status: string | undefined;
      StatusCode: number | undefined;
    }
  | undefined
> {
  const region = event.ResourceProperties['region'];
  const adminAccountId = event.ResourceProperties['adminAccountId'];
  const kmsKeyArn: string | undefined = event.ResourceProperties['kmsKeyArn'] ?? undefined;
  const solutionId = event.ResourceProperties['solutionId'];

  const auditManagerClient = new AWS.AuditManager({ region: region, customUserAgent: solutionId });

  const auditManagerAdminAccount = await isAuditManagerEnable(auditManagerClient);

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      if (auditManagerAdminAccount.accountId === adminAccountId) {
        console.warn(
          `GuardDuty admin account ${auditManagerAdminAccount.accountId} is already an admin account, in ${region} region. No action needed`,
        );
        return { Status: 'Success', StatusCode: 200 };
      } else {
        console.log(
          `Started enableOrganizationAdminAccount function in ${event.ResourceProperties['region']} region for account ${adminAccountId}`,
        );
        await throttlingBackOff(() =>
          auditManagerClient.registerAccount({ delegatedAdminAccount: adminAccountId, kmsKey: kmsKeyArn }).promise(),
        );
        await throttlingBackOff(() =>
          auditManagerClient.registerOrganizationAdminAccount({ adminAccountId: adminAccountId }).promise(),
        );
      }

      return { Status: 'Success', StatusCode: 200 };

    case 'Delete':
      console.log(auditManagerAdminAccount.accountId);
      if (auditManagerAdminAccount.accountId) {
        if (auditManagerAdminAccount.accountId === adminAccountId) {
          console.log(
            `Started disableOrganizationAdminAccount function in ${event.ResourceProperties['region']} region for account ${adminAccountId}`,
          );
          await throttlingBackOff(() =>
            auditManagerClient
              .deregisterOrganizationAdminAccount({
                adminAccountId: adminAccountId,
              })
              .promise(),
          );
          await throttlingBackOff(() => auditManagerClient.deregisterAccount().promise());
        }
      }

      return { Status: 'Success', StatusCode: 200 };
  }
}

async function isAuditManagerEnable(auditManagerClient: AWS.AuditManager): Promise<{ accountId: string | undefined }> {
  try {
    const response = await throttlingBackOff(() => auditManagerClient.getOrganizationAdminAccount().promise());
    const adminAccount = response.adminAccountId;

    if (adminAccount === undefined) {
      return { accountId: undefined };
    }
    return { accountId: adminAccount };
    //eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    if (error.code === 'AccessDeniedException') {
      console.log('Admin account not enabled');
      return { accountId: undefined };
    } else {
      throw error;
    }
  }
}
