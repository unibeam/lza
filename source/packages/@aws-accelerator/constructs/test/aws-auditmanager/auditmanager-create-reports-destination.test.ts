/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { AuditManagerDefaultReportsDestination } from '../../lib/aws-auditmanager/auditmanager-reports-destination';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(AuditManagerDefaultReportsDestination): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

new AuditManagerDefaultReportsDestination(stack, 'AuditManagerDefaultReportsDestination', {
  bucket: `s3//aws-accelerator-auditmgr-${stack.account}-${stack.region}`,
  defaultReportsDestinationType: 'S3',
  bucketKmsKey: new cdk.aws_kms.Key(stack, 'BucketKey', {}),
  kmsKey: new cdk.aws_kms.Key(stack, 'CustomKey', {}),
  logRetentionInDays: 3653,
});

/**
 * AuditManagerDefaultReportsDestination construct test
 */
describe('AuditManagerDefaultReportsDestination', () => {
  snapShotTest(testNamePrefix, stack);
});
