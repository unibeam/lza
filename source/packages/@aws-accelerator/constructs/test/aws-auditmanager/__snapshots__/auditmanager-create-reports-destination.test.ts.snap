// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditManagerDefaultReportsDestination Construct(AuditManagerDefaultReportsDestination):  Snapshot Test 1`] = `
{
  "Resources": {
    "AuditManagerDefaultReportsDestinationAFD20D60": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderLogGroupF5AC3566",
      ],
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderHandler6BCBC433",
            "Arn",
          ],
        },
        "bucket": {
          "Fn::Join": [
            "",
            [
              "s3//aws-accelerator-auditmgr-",
              {
                "Ref": "AWS::AccountId",
              },
              "-",
              {
                "Ref": "AWS::Region",
              },
            ],
          ],
        },
        "defaultReportsDestinationType": "S3",
        "kmsKeyArn": {
          "Fn::GetAtt": [
            "BucketKey7092080A",
            "Arn",
          ],
        },
        "region": {
          "Ref": "AWS::Region",
        },
      },
      "Type": "Custom::AuditManagerCreateDefaultReportsDestination",
      "UpdateReplacePolicy": "Delete",
    },
    "BucketKey7092080A": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderHandler6BCBC433": {
      "DependsOn": [
        "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderRoleAEE72AE5",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderRoleAEE72AE5",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderLogGroupF5AC3566": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "CustomKey1E6D0D07",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderHandler6BCBC433",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomAuditManagerCreateDefaultReportsDestinationCustomResourceProviderRoleAEE72AE5": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "auditmanager:UpdateSettings",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                  "Sid": "AuditManagerCreatePublishingDestinationCommandTaskAuditManagerActions",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomKey1E6D0D07": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
